### PR TITLE
Fix or better handle all async loops in the solution [Part 2 - Wallet Features]

### PR DESCRIPTION
**Pull request by fassadlr**
Refactor Wallet Features to implement Stop and Start methods to dispose of async loops IF they have been set or used.

### CODE REVIEW

File: Stratis.Bitcoin.Features.LightWallet/LightWalletSyncManager.cs [hunk 1]
```diff
@@ -198,12 +209,10 @@ public void SyncFrom(int height)
         private void StartSync(int height)
         {
             // TODO add support for the case where there is a reorg, like in the initialize method
-            var chainedBlock = this.chain.GetBlock(height);
-            if (chainedBlock == null)
-                throw new WalletException("Invalid block height");
-            this.walletTip = chainedBlock;
+            ChainedBlock chainedBlock = this.chain.GetBlock(height);
+            this.walletTip = chainedBlock ?? throw new WalletException("Invalid block height");
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.LightWallet/LightWalletSyncManager.cs]: Hmm nice

File: Stratis.Bitcoin.Features.Wallet/IWalletSyncManager.cs [hunk 2]
```diff
@@ -1,46 +1,53 @@
...
 
-		/// <summary>
-		/// Processes a new block
-		/// </summary>
-		/// <param name="block"></param>
-		void ProcessBlock(Block block);
+        /// <summary>
+        /// Stops the walletSyncManager.
+        /// <para>
+        /// We need to call <see cref="Stop"/> explicitly to check that the internal async cloop isn't still running
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/IWalletSyncManager.cs]: typo 'cloop '

File: Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 3]
```diff
@@ -112,23 +117,36 @@ public void Initialize()
...
+        }
+
+        /// <inheritdoc />
+        public void Stop()
+        {
+            if (this.asyncLoop != null)
+                this.asyncLoop.Dispose();
+
+            // safely persist the wallets to the file system before disposing
+            foreach (var wallet in this.Wallets)
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Can you not just call the single method `  this.SaveToFile();` same as in the loop?

File: Stratis.Bitcoin.Features.Wallet/WalletFeature.cs [hunk 4]
```diff
@@ -63,20 +63,20 @@ public void AddNodeStats(StringBuilder benchLogs)
...
         /// <inheritdoc />
         public override void Stop()
         {
-            this.blockSubscriberdDisposable.Dispose();
-            this.transactionSubscriberdDisposable.Dispose();
+            this.blockSubscriberDisposable.Dispose();
+            this.transactionSubscriberDisposable.Dispose();
 
-            this.walletManager.Dispose();
+            this.walletManager.Stop();
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/WalletFeature.cs]: I think you forgot
```  this.walletSyncManager.Stop();```

File: Stratis.Bitcoin.Features.Wallet/WalletFeePolicy.cs [hunk 5]
```diff
@@ -4,13 +4,13 @@
...
 using Stratis.Bitcoin.Utilities;
 using System;
-using System.Threading.Tasks;
 
 namespace Stratis.Bitcoin.Features.Wallet
 {
     public interface IWalletFeePolicy
     {
-        Task Initialize();
+        void Start();
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/WalletFeePolicy.cs]: We could consider creating an interface 
```
IStratable
{
   Strat()
   Stop()
}
```

File: Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs [hunk 6]
```diff
@@ -66,8 +64,12 @@ public virtual Task Initialize()
                 this.walletManager.WalletTipHash = fork.HashBlock;
                 this.walletTip = fork;
             }
+        }
 
-            return Task.CompletedTask;
+        /// <inheritdoc />
+        public void Stop()
+        {
+            return;
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: Return is not needed.

File: Stratis.Bitcoin.Features.Wallet/IWalletSyncManager.cs (hunk 2 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Wallet/IWalletSyncManager.cs]: Fixed

File: Stratis.Bitcoin.Features.LightWallet/LightWalletSyncManager.cs (hunk 1 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.LightWallet/LightWalletSyncManager.cs]: 👍

File: Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 3 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: You are correct, we can just call this.SaveToFile(). I renamed them to SaveWallets and SaveWallet.. its more descriptive ;)

File: Stratis.Bitcoin.Features.Wallet/WalletFeature.cs (hunk 4 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Wallet/WalletFeature.cs]: Fixed

File: Stratis.Bitcoin.Features.Wallet/WalletFeePolicy.cs (hunk 5 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Wallet/WalletFeePolicy.cs]: That was where I was heading ;) But lets do this refactor in a separate task? I want to standardize all of the feature services that have Start and Stop methods.

File: Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs (hunk 6 above)
Comment (fassadlr) [reply Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: Fixed

File: Stratis.Bitcoin.Features.Wallet/WalletFeePolicy.cs (hunk 5 above)
Comment (dangershony) [reply Stratis.Bitcoin.Features.Wallet/WalletFeePolicy.cs]: Create an issue (and task on the board) so we have it in the back log.

**Comment by dangershony** [comment]
Ah FYI this test failed `NotifyWithSyncFromSetBroadcastsOnNextBlock`

**Pull request merged**